File: June-LeetCoding-Challenge/28-Reconstruct-Itinerary/Reconstruct-Itinerary.cs
public class Solution
{
    public IList<string> FindItinerary(IList<IList<string>> tickets)
    {
        var result = new List<string>();
        var dict = new Dictionary<string, List<string>>();
        for (int i = 0; i < tickets.Count(); i++)
        {
            var from = tickets[i][0];
            var to = tickets[i][1];
            if (!dict.ContainsKey(from))
            {
                dict[from] = new List<string>();
            }
            dict[from].Add(to);
        }
        //Sort items in value list for each dictionary key
        foreach (var kv in dict)
        {
            kv.Value.Sort();
        }
        var stack = new Stack<string>();
        stack.Push("JFK");
        while (stack.Count > 0)
        {
            while (dict.ContainsKey(stack.Peek()) && dict[stack.Peek()].Count > 0)
            {
                var firstDest = dict[stack.Peek()].First();
                dict[stack.Peek()].Remove(firstDest);
                stack.Push(firstDest);
            }
            result.Insert(0, stack.Pop());
        }
        return result;
    }
}

File: June-LeetCoding-Challenge/28-Reconstruct-Itinerary/Reconstruct-Itinerary.cpp
class Solution {
public:
    int n;
    vector<string> ans;
    map<string, vector<string>> a;
    
    
    bool dfs(string s){
        ans.push_back(s);
        if (ans.size() == n + 1) return true;
        for (auto& t: a[s]){
            if (t != "") {
                string tmp = t;
                t = "";
                if (dfs(tmp)) return true;
                t = tmp;
            }   
        }
        ans.pop_back();
        return 0;
    }
    
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        n = tickets.size();
        
        for (int i = 0; i < n; i++)
            a[tickets[i][0]].push_back(tickets[i][1]);
        
        for (int i = 0; i < n; i++)
            sort(a[tickets[i][0]].begin(), a[tickets[i][0]].end());
        dfs("JFK");
        return ans;
    }
};